File: 0054. Spiral Matrix I and 59 Sprial Matrix II/Solution0054.cpp
#include <vector>

using namespace std;

class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        vector<int>res;
        int row = matrix.size(), col = matrix[0].size();
        int topRow = 0, bottomRow = row - 1;
        int leftCol = 0, rightCol = col - 1;        
        while (topRow < bottomRow && leftCol < rightCol) {
            for (int c = leftCol; c < rightCol; c++) {
                res.push_back(matrix[topRow][c]);
            }
            for (int r = topRow; r < bottomRow; r++) {
                res.push_back(matrix[r][rightCol]);
            }
            for (int c = rightCol; c > leftCol; c--) {
                res.push_back(matrix[bottomRow][c]);
            }
            for (int r = bottomRow; r > topRow; r--) {
                res.push_back(matrix[r][leftCol]);
            }
            topRow++; bottomRow--;
            leftCol++; rightCol--;             
        }
        if (topRow == bottomRow) {
            while (leftCol <= rightCol) {
                res.push_back(matrix[bottomRow][leftCol++]);
            }
        } else if(leftCol == rightCol) {
            while(topRow <= bottomRow) {
                res.push_back(matrix[topRow++][leftCol]); // !!!
            }
        }  
        
        return res;        
    }
};

File: 0054. Spiral Matrix I and 59 Sprial Matrix II/Solution0054.py
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        res = []
        top_row, bottom_row = 0, len(matrix) - 1         
        left_col, right_col = 0, len(matrix[0]) - 1
        while (top_row < bottom_row and left_col < right_col):
            for i in range(left_col, right_col):
                res.append(matrix[top_row][i])
            for i in range(top_row, bottom_row):
                res.append(matrix[i][right_col])
            for i in range(right_col,left_col,-1):
                res.append(matrix[bottom_row][i])
            for i in range(bottom_row,top_row,-1):
                res.append(matrix[i][left_col])
            top_row += 1
            bottom_row -= 1
            left_col += 1
            right_col -= 1
         
        if top_row == bottom_row:
            while left_col <= right_col:
                res.append(matrix[top_row][left_col])
                left_col += 1 #-
        elif left_col == right_col:
            while top_row <= bottom_row:
                res.append(matrix[top_row][left_col])
                top_row += 1 #|
        return res